test_rclcpp/test/test_executor.cpp

```diff
@@ -40,6 +40,7 @@
 using namespace std::chrono_literals;
 
 TEST(CLASSNAME(test_executor, RMW_IMPLEMENTATION), recursive_spin_call) {
+  if (!rclcpp::ok()) {rclcpp::init(0, nullptr);}
```

Comment:
Why this line instead of `init()` and `shutdown()` in `main()`? It looks like `rclcpp` won't be shut down.

Reply:
Calling shutdown explicitly is not required.

A few of the test files with the init/shutdown in the main actually had one or more tests that called shutdown themselves, causing tests that ran after them to fail because "init had not been called" (since the shutdown in the other test).

I found this to be cleaner, as it ensures init has been called without shutting down and initializing again unnecessarily. Also it is resilient to tests being filtered by gtest.

Reply:
> Calling shutdown explicitly is not required.

To be clear, you may call shutdown manually, or it is called in the SIGINT signal handler for each Context which elected to have `shutdown_on_sigint=true` (the default) and which are not already shutdown, or finally it is called in the destructor of Context right before finalizing the rcl context object, if not already shutdown.